File: monsoon3.0/src/main/java/wtf/monsoon/impl/module/combat/Velocity.java
package wtf.monsoon.impl.module.combat;

import com.mojang.authlib.GameProfile;
import io.github.nevalackin.homoBus.Listener;
import io.github.nevalackin.homoBus.annotations.EventLink;
import kotlin.jvm.internal.Intrinsics;
import net.minecraft.client.entity.EntityOtherPlayerMP;
import net.minecraft.network.play.client.C02PacketUseEntity;
import net.minecraft.network.play.client.C0APacketAnimation;
import net.minecraft.network.play.server.S12PacketEntityVelocity;
import net.minecraft.network.play.server.S27PacketExplosion;
import wtf.monsoon.Wrapper;
import wtf.monsoon.api.module.Category;
import wtf.monsoon.api.module.Module;
import wtf.monsoon.api.setting.Setting;
import wtf.monsoon.api.util.entity.PlayerUtil;
import wtf.monsoon.api.util.misc.PacketUtil;
import wtf.monsoon.api.util.misc.StringUtil;
import wtf.monsoon.impl.event.EventPacket;
import wtf.monsoon.impl.event.EventUpdate;
import wtf.monsoon.impl.module.movement.Speed;

import java.util.Objects;
import java.util.UUID;

public class Velocity extends Module {

    private final Setting<Mode> mode = new Setting<>("Mode", Mode.CANCEL)
            .describedBy("The mode of velocity.");

    private final Setting<Integer> horMod = new Setting<>("Horizontal", 0)
            .minimum(0)
            .maximum(100)
            .incrementation(1)
            .describedBy("Horizontal velocity modifier.")
            .visibleWhen(() -> mode.getValue().equals(Mode.CANCEL));

    private final Setting<Integer> vertMod = new Setting<>("Vertical", 0)
            .minimum(0)
            .maximum(100)
            .incrementation(1)
            .describedBy("Vertical velocity modifier.")
            .visibleWhen(() -> mode.getValue().equals(Mode.CANCEL));

    private final Setting<Double> strength = new Setting<>("Strength", 3D)
            .minimum(1D)
            .maximum(10D)
            .incrementation(1D)
            .describedBy("The strength of velocity.")
            .visibleWhen(() -> mode.getValue().equals(Mode.DRAG_CLICK));

    private boolean receivedVelocity;

    public Velocity() {
        super("Velocity", "Take no knockback.", Category.COMBAT);

        this.setMetadata(() -> {
            switch (mode.getValue()) {
                case DRAG_CLICK: return StringUtil.formatEnum(mode.getValue()) + ", " + strength.getValue();
                case WATCHDOG: return "Watchdog";
                case CANCEL: return horMod.getValue() + "% " + vertMod.getValue() + "%";
            }

            return "";
        });
    }

    @Override
    public void onDisable() {
        super.onDisable();
        receivedVelocity = false;
    }

    @EventLink
    public final Listener<EventPacket> eventPacketListener = e -> {
        if (mc.thePlayer == null || mc.theWorld == null) {
            return;
        }

        if (e.getPacket() instanceof S12PacketEntityVelocity) {
            S12PacketEntityVelocity packet = (S12PacketEntityVelocity) e.getPacket();

            if (packet.getEntityID() == mc.thePlayer.getEntityId()) {
                int amount = ((packet.getMotionX() + packet.getMotionY() + packet.getMotionZ()) / 3) / 200;
                Wrapper.getSexToyManager().vibrate(amount);
            }

            if (packet.getEntityID() == mc.thePlayer.getEntityId()) {
                receivedVelocity = true;
            }

            switch (mode.getValue()) {
                case CANCEL:
                    if(horMod.getValue() == 0 && vertMod.getValue() == 0) e.setCancelled(true);

                    packet.setMotionX((int) ((packet.getMotionX() * 0.01) * horMod.getValue()));
                    packet.setMotionY((int) ((packet.getMotionY() * 0.01) * vertMod.getValue()));
                    packet.setMotionZ((int) ((packet.getMotionZ() * 0.01) * horMod.getValue()));
                    break;
                case WATCHDOG:

                    if(!(Wrapper.getModule(Speed.class).isEnabled() && Wrapper.getModule(Speed.class).getMode().equals(Mode.WATCHDOG))) {
                        if (packet.getEntityID() == mc.thePlayer.getEntityId()) {
                            e.setCancelled(true);
                            // mc.thePlayer.motionY = packet.getMotionY() / 8000.0D;
                            mc.thePlayer.motionY = packet.getMotionY() / 8000.0D;
                        }
                    }

                    break;
            }
        } else if (e.getPacket() instanceof S27PacketExplosion) {
            switch (mode.getValue()) {
                case CANCEL:
                    e.setCancelled(true);
                    break;

                case WATCHDOG:
                    if(mc.thePlayer.hurtTime > 0) {
                        e.setCancelled(true);
                        mc.thePlayer.motionY += 0.001 - Math.random() / 100f;
                    }

                    break;
            }
        }
    };

    @EventLink
    public final Listener<EventUpdate> eventUpdateListener = event -> {
        if (mode.getValue() == Mode.DRAG_CLICK) {
            if (mc.thePlayer.hurtTime == 10 && receivedVelocity) {
                for (int i = 0; i < 10; i++) {
                    EntityOtherPlayerMP fakePlayer = new EntityOtherPlayerMP(mc.theWorld, new GameProfile(UUID.randomUUID(), "a"));

                    PacketUtil.sendPacketNoEvent(new C0APacketAnimation());
                    PacketUtil.sendPacketNoEvent(new C02PacketUseEntity(fakePlayer, C02PacketUseEntity.Action.ATTACK));
                    mc.thePlayer.motionX *= (1. / strength.getValue());
                    mc.thePlayer.motionZ *= (1. / strength.getValue());
                }
            }
        }

        if (mc.thePlayer.hurtTime == 0) {
            receivedVelocity = false;
        }
    };

    enum Mode {
        CANCEL, WATCHDOG, DRAG_CLICK
    }

}

File: monsoon-3.x-v2/src/main/java/wtf/monsoon/client/modules/combat/Velocity.kt
package wtf.monsoon.client.modules.combat

import com.mojang.authlib.GameProfile
import me.bush.eventbuskotlin.EventListener
import me.bush.eventbuskotlin.listener
import net.minecraft.client.entity.EntityOtherPlayerMP
import net.minecraft.network.play.client.C02PacketUseEntity
import net.minecraft.network.play.client.C0APacketAnimation
import net.minecraft.network.play.server.S12PacketEntityVelocity
import net.minecraft.network.play.server.S27PacketExplosion
import wtf.monsoon.backend.Category
import wtf.monsoon.backend.module.Module
import wtf.monsoon.backend.setting.Setting
import wtf.monsoon.client.event.EventPacket
import wtf.monsoon.client.event.EventPreMotion
import wtf.monsoon.client.event.EventUpdate
import wtf.monsoon.client.util.misc.StringUtil
import wtf.monsoon.client.util.network.PacketUtil
import java.util.*

class Velocity : Module("Velocity", "Take reduced or zero knockback.", Category.COMBAT) {

    private val mode: Setting<Mode> = Setting("Mode", "The mode of velocity.", Mode.CANCEL)

    private val horMod: Setting<Int> = Setting("Horizontal", "Horizontal velocity modifier.", 0)
        .minimum(0)
        .maximum(100)
        .incrementation(1)
        .visibleWhen { mode.getValue() == Mode.CANCEL }

    private val vertMod: Setting<Int> = Setting("Vertical", "Vertical velocity modifier.", 0)
        .minimum(0)
        .maximum(100)
        .incrementation(1)
        .visibleWhen { mode.getValue() == Mode.CANCEL }

    private val strength: Setting<Double> = Setting("Strength", "The strength of velocity.", 3.0)
        .minimum(1.0)
        .maximum(10.0)
        .incrementation(1.0)
        .visibleWhen { mode.getValue() == Mode.DRAG_CLICK }

    private var receivedVelocity = false

    override var metadata: () -> String = {
        when (mode.getValue()) {
            Mode.DRAG_CLICK -> "${StringUtil.formatEnum(mode.getValue())}, ${strength.getValue()}"
            Mode.WATCHDOG -> "Watchdog"
            Mode.CANCEL -> "${horMod.getValue()}% ${vertMod.getValue()}%"
            else -> ""
        }
    }

    override fun disable() {
        super.disable()
        receivedVelocity = false
    }

    @EventListener
    val packet = fun(it: EventPacket) {
        if (mc.thePlayer == null || mc.theWorld == null) {
            return
        }

        if (it.packet is S12PacketEntityVelocity) {
            val packet = it.packet as S12PacketEntityVelocity

            if (packet.entityID == mc.thePlayer.entityId) {
                receivedVelocity = true
            }
            when (mode.getValue()) {
                Mode.CANCEL -> {
                    if (horMod.getValue() === 0 && vertMod.getValue() === 0) it.cancel()
                    packet.motionX = (packet.motionX * 0.01 * horMod.getValue()).toInt()
                    packet.motionY = (packet.motionY * 0.01 * vertMod.getValue()).toInt()
                    packet.motionZ = (packet.motionZ * 0.01 * horMod.getValue()).toInt()
                }

                Mode.WATCHDOG -> {
                    // if (!(Wrapper.getModule(Speed::class.java).isEnabled() && Wrapper.getModule(Speed::class.java).getMode().equals(Mode.WATCHDOG))) {
                        if (packet.entityID == mc.thePlayer.entityId) {
                            it.cancel()
                            mc.thePlayer.motionY = packet.motionY / 8000.0
                        }
                    // }
                }
                else -> {}
            }
        } else if (it.packet is S27PacketExplosion) {
            when (mode.getValue()) {
                Mode.CANCEL -> it.cancel()
                Mode.WATCHDOG -> if (mc.thePlayer.hurtTime > 0) {
                    it.cancel()
                    mc.thePlayer.motionY += 0.001 - Math.random() / 100f
                }
                else -> {}
            }
        }
    }

    @EventListener
    val preMotion = fun(_: EventPreMotion) {
        if (mode.getValue() === Mode.DRAG_CLICK) {
            if (mc.thePlayer.hurtTime === 10 && receivedVelocity) {
                for (i in 0..9) {
                    val fakePlayer = EntityOtherPlayerMP(mc.theWorld, GameProfile(UUID.randomUUID(), "a"))
                    PacketUtil.sendPacketNoEvent(C0APacketAnimation())
                    PacketUtil.sendPacketNoEvent(C02PacketUseEntity(fakePlayer, C02PacketUseEntity.Action.ATTACK))
                    mc.thePlayer.motionX *= 1.0 / strength.getValue()
                    mc.thePlayer.motionZ *= 1.0 / strength.getValue()
                }
            }
        }

        if (mc.thePlayer.hurtTime === 0) {
            receivedVelocity = false
        }
    }

    internal enum class Mode {
        CANCEL, WATCHDOG, DRAG_CLICK
    }
}